### PR TITLE
Speed up the use of DDFilteredViews

### PR DESCRIPTION
**Pull request by Dr15Jones**
Profiling of multi-threaded GEN+SIM jobs showed that the initialization time in RunManagerMTWorker::initializeThread was severely impacting the total event throughput for jobs with large numbers of streams (>100) and only modest numbers of events per stream. Using 12 cores on a standard Xeon system under CMSSW_9_0_0_pre5 the call to initializeThread took 517 seconds. After all the changes in this pull request it now takes 51 seconds.

The changes were
-Replace DDSpecificsFilter with filters that are specifics to the needs of the query.
-Have DDFilteredView only take one DDFilter (no code used more than one every anyway).
-Test DDFilters with a standard unit test.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
@ianna I'd appreciate it if you'd do a careful code review of this pull request.

**Comment by Dr15Jones** [comment]
@ianna is there any reason to keep DetectorDescription/RegressionTest/test/tutorial.cc? It looks like it only works if someone externally fills the singletons for a DDCompactView.

File: Geometry/CSCGeometryBuilder/src/CSCGeometryParsFromDD.cc [hunk 1]
```diff
@@ -34,14 +34,10 @@ bool CSCGeometryParsFromDD::build( const DDCompactView* cview
 
   DDSpecificsFilter filter;
```
Comment (Dr15Jones) [inline Geometry/CSCGeometryBuilder/src/CSCGeometryParsFromDD.cc]: I should have changed this to `DDSpecificsMatchesValueFilter`.

File: Geometry/CaloEventSetup/interface/CaloGeometryLoader.icc [hunk 2]
```diff
@@ -42,18 +42,12 @@ CaloGeometryLoader<T>::CaloGeometryLoader()
    m_filter.setCriteria( DDValue( "SensitiveDetector",
```
Comment (Dr15Jones) [inline Geometry/CaloEventSetup/interface/CaloGeometryLoader.icc]: m_filter should be changed to be a `DDAndFilter<>`

File: Geometry/EcalTestBeam/plugins/EcalTBHodoscopeGeometryLoaderFromDDD.cc [hunk 3]
```diff
@@ -136,17 +135,11 @@ DDFilter* EcalTBHodoscopeGeometryLoaderFromDDD::getDDFilter()
    filter->setCriteria( DDValue( "SensitiveDetector",
```
Comment (Dr15Jones) [inline Geometry/EcalTestBeam/plugins/EcalTBHodoscopeGeometryLoaderFromDDD.cc]: `filter` should be converted to an `DDAndFilter`.

**Comment by ianna** [comment]
@Dr15Jones - very impressive, thanks! I think, tutorial.cc can be safely removed.

**Comment by Dr15Jones** [comment]
please test

**Comment by ianna** [comment]
1001.0 workflow fails due to:
```
Error in <TFile::TFile>: file /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/files/1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5/17778-DQM_V0001_R000165121__Express__PCLTest__ALCAPROMPT.root does not exist
Error in <TH1F::ComputeIntegral>: Integral = zero
Reading meta from commandline
Analysing Histograms located in directory  at: 
 o /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/files/1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5/CMSSW_9_1_X_2017-03-05-1100-DQM_V0001_R000165121__Express__PCLTest__ALCAPROMPT.root
 o /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/files/1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5/17778-DQM_V0001_R000165121__Express__PCLTest__ALCAPROMPT.root
Directory DQMData does not exist: skipping. Is this a custom rootfile?
++++++++++++++++++++++++++++++
Output Directory will be  /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/default-comparison/1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5

Rundir not there: Is this a generic rootfile?

Setting run number to 0
Finished

 - summary of 0 tests:
 o Failiures: 0.00% (0/0)
 o Nulls: 0.00% (0/0) 
 o Successes: 1.00% (0/0) 
 o Skipped: 0.00% (0/0) 
 o Missing objects: 0
Pickleing the directory as /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/default-comparison/1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5.pkl in dir /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/default-comparison/1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5
Reading directory from /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/default-comparison/1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5.pkl
Calculating stats for the directory...
Producing html...
Traceback (most recent call last):
  File "/cvmfs/cms-ib.cern.ch/nweek-02462/slc6_amd64_gcc530/cms/cmssw/CMSSW_9_1_X_2017-03-05-0000/bin/slc6_amd64_gcc530/compare_using_files.py", line 345, in <module>
    directory2html(directory, options.hash_name, options.standalone)
  File "/cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_1_X_2017-03-05-1100/python/Utilities/RelMon/directories2html.py", line 464, in directory2html
    page_html+=get_rank_section(directory)
  File "/cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_1_X_2017-03-05-1100/python/Utilities/RelMon/directories2html.py", line 390, in get_rank_section
    scale = gPad.GetUymax()/rightmax
ZeroDivisionError: float division by zero
```

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
As far as I can tell, all the comparisons are normal.

**Comment by ianna** [comment]
@Dr15Jones - Yes, I'm just running some extra tests.

**Review by ianna — APPROVED** [review]
The following files can be removed:
DetectorDescription/RegressionTest/test/tutorial.cc
DetectorDescription/Core/interface/DDQuery.h
DetectorDescription/Core/src/DDQuery.cc

**Comment by ianna** [comment]
+1

**Comment by kpedro88** [comment]
+1

**Comment by civanch** [comment]
For some reasons my local comparisons show differences for muon hits (nothing for calorimeters) but official comparison are OK. May this be due to some change in rotations?

Another concern: there is a report about broken CMS code rules in this PR (?) - this is unusual.

**Comment by Dr15Jones** [comment]
@civanch the coding rule violations are ones already in the IB. From this pull request the violations are:
```
/CalibMuon/DTCalibration/plugins/DTGeometryParserFromDDD.cc
[34]
/Geometry/EcalTestBeam/plugins/EcalTBGeometryBuilder.cc
[60, 66]
/Geometry/VeryForwardGeometryBuilder/src/RPAlignmentCorrectionsMethods.cc
[63]
/Geometry/VeryForwardGeometryBuilder/src/RPAlignmentCorrectionsDataSequence.cc
[98]
/Geometry/VeryForwardGeometryBuilder/plugins/TotemRPGeometryESModule.cc
[181, 212]
```

If you look at the most recent IB Q&A you'll see those same items:
https://cmssdt.cern.ch/SDT/html//rc//slc6_amd64_gcc530/www/mon/9.1-mon-23/CMSSW_9_1_X_2017-03-06-2300//codeRules/cmsCRPage.html

**Comment by Dr15Jones** [comment]
@civanch How can one reproduce the comparison problems you are seeing locally?

**Comment by civanch** [comment]
@Dr15Jones , I run 600 MeanBias events at recent IB with 8 threads, then merge locally this PR and run again. Number of hits in CSC is not big for this statistics (~500) but the difference is visible. It may be a false alarm, because these hits may be due to few events. My little worry is due to expectation that nothing should be affected and sim history should not change at all. Probably, I have to rerun with 1 thread.

In calorimeters number of hits is significant and distributions are really very similar.

**Comment by Dr15Jones** [comment]
@civanch do you happen to have the exact cmsDriver command you used to generate the configuration?

**Comment by civanch** [comment]
+1
when I run with 1 thread (or stream?) the hits are identical. The false alarm was connected with 8-thread runs.

**Comment by Dr15Jones** [comment]
@civanch David Dagenhart and I discussed this and one would expect the possibility of different results using multiple threads. We tie random number seeds to StreamIDs. However, when running different jobs using the same configuration with multiple threads, the streams in the two different jobs are likely to process different number of events and therefore use different random number seeds.

As an example, say you are going to generate 8 events and use 4 streams. In the first job you might get the following stream to #events processed
````
stream 1: 2 events
stream 2: 2 events
stream 3: 2 events
stream 4: 2 events
```
in another job you could get
```
stream 1: 1 event
stream 2: 1 event
stream 3: 3 events
stream 4: 3 events
```
So between these two jobs, you'd have 6 events which saw the same random numbers and 2 events which saw different random numbers.

**Comment by Dr15Jones** [comment]
@dmitrijus @vanbesien @cerminar @franzoni @mmusich @ghellwig @arunhep please review this pull request.

**Comment by dmitrijus** [comment]
+1

**Comment by Dr15Jones** [comment]
@cerminar @franzoni @mmusich @ghellwig @arunhep please review. alca is the last signature needed.

**Pull request merged**